google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py

```diff
@@ -0,0 +1,162 @@
+# limitations under the License.
+
+import sys
+from datetime import datetime
+import warnings
+from typing import Optional
+
+from opentelemetry import trace
+from opentelemetry.trace.propagation import set_span_in_context
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+from google.pubsub_v1 import types as gapic_types
+from google.cloud.pubsub_v1.open_telemetry.context_propagation import (
+    OpenTelemetryContextSetter,
+)
+
+
+class PublishMessageWrapper:
+    _OPEN_TELEMETRY_TRACER_NAME: str = "google.cloud.pubsub_v1.publisher"
+    _OPEN_TELEMETRY_MESSAGING_SYSTEM: str = "gcp_pubsub"
+    _OPEN_TELEMETRY_PUBLISHER_BATCHING = "publisher batching"
+
+    _PUBLISH_START_EVENT: str = "publish start"
+    _PUBLISH_FLOW_CONTROL: str = "publisher flow control"
+
+    def __init__(self, message: gapic_types.PubsubMessage):
+        self._message: gapic_types.PubsubMessage = message
+
+    @property
+    def message(self):
+        return self._message
+
+    @message.setter  # type: ignore[no-redef]  # resetting message value is intentional here
+    def message(self, message: gapic_types.PubsubMessage):
+        self._message = message
+
+    @property
+    def create_span(self):
+        return self._create_span
+
+    def __eq__(self, other):  # pragma: NO COVER
+        """Used for pytest asserts to compare two PublishMessageWrapper objects with the same message"""
+        if isinstance(self, other.__class__):
+            return self.message == other.message
+        return False
+
+    def start_create_span(self, topic: str, ordering_key: str) -> None:
+        tracer = trace.get_tracer(self._OPEN_TELEMETRY_TRACER_NAME)
+        topic_short_name = topic.split("/")[3]
+        with tracer.start_as_current_span(
+            name=f"{topic_short_name} create",
+            attributes={
+                "messaging.system": self._OPEN_TELEMETRY_MESSAGING_SYSTEM,
+                "messaging.destination.name": topic_short_name,
+                "code.function": "publish",
+                "messaging.gcp_pubsub.message.ordering_key": ordering_key,
+                "messaging.operation": "create",
+                "gcp.project_id": topic.split("/")[1],
+                "messaging.message.body.size": sys.getsizeof(self._message.data),
+            },
+            kind=trace.SpanKind.PRODUCER,
+            end_on_exit=False,
+        ) as create_span:
+            create_span.add_event(
+                name=self._PUBLISH_START_EVENT,
+                attributes={
+                    "timestamp": str(datetime.now()),
+                },
+            )
+            self._create_span: trace.Span = create_span
+            TraceContextTextMapPropagator().inject(
+                carrier=self._message,
+                setter=OpenTelemetryContextSetter(),
+            )
+
+    def end_create_span(self, exc: Optional[BaseException] = None) -> None:
+        if self._create_span is None:  # pragma: NO COVER
```

Comment:
The tests already exercise the expected code path, i.e if we are calling end_create_span on the publish wrapper, the code must have started it previously. If it hasn't been started, then it is an error in the code. The pragma NO Cover is for test coverage for the unexpected case of the create_span being None, when its not supposed to. If the pragma no cover is replaced with a test, the test would have to be mocking this scenario since it would not occur / isn't currently occurring in the code. 

But I have created a bug to just evaluate usage of pragma no cover across the repo here: https://github.com/googleapis/python-pubsub/issues/1249

Reply:
> I'd just end up mocking the wrapper to create error scenarios where end_create_span is called when there is no span to end.

Please add a TODO to add these tests as part of follow on work

Reply:
I just ended up adding the tests with the unexpected scenarios themselves (which was what I would have had to do as part of the TODO)

Reply:
Resolving this comment since I've addressed it. But please feel free to reopen if anything additional needs to be done here.